### PR TITLE
BLD: added cython as a setup dependency

### LINKED ISSUE #7959: numpy's setup depends on cython, but it does not declare this.
when installing numpy with
`easy_install --user https://github.com/numpy/numpy/archive/master.zip`
I get the error:

```
NFO: Running setup.py -q bdist_egg --dist-dir /tmp/easy_install-IeB3Vc/numpy.git@numpysetup/egg-dist-tmp-ueY6qq
/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/setup.py:331: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates
  warnings.warn("Unrecognized setuptools command, proceeding with "
Running from numpy source directory.
Cythonizing sources
Processing numpy/random/mtrand/mtrand.pyx
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named Cython.Compiler.Main
Traceback (most recent call last):
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/tools/cythonize.py", line 199, in <module>
    main()
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/tools/cythonize.py", line 195, in main
    find_process_files(root_dir)
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/tools/cythonize.py", line 187, in find_process_files
    process(cur_dir, fromfile, tofile, function, hash_db)
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/tools/cythonize.py", line 161, in process
    processor_function(fromfile, tofile)
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/tools/cythonize.py", line 81, in process_pyx
    raise Exception('Cython failed')
Exception: Cython failed
Traceback (most recent call last):
  File "setup.py", line 76, in <module>
    action_target(PACKAGE)
  File "/home/jens/.local/lib/python2.7/site-packages/vsc_install-0.10.10-py2.7.egg/vsc/install/shared_setup.py", line 1418, in action_target
    _fvs('action_target function')().action_target(package, *args, **kwargs)
  File "/home/jens/.local/lib/python2.7/site-packages/vsc_install-0.10.10-py2.7.egg/vsc/install/shared_setup.py", line 1406, in action_target
    setupfn(**x)
  File "/usr/lib64/python2.7/distutils/core.py", line 111, in setup
    _setup_distribution = dist = klass(attrs)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/dist.py", line 348, in __init__
    self.fetch_build_eggs(attrs['setup_requires'])
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/dist.py", line 394, in fetch_build_eggs
    replace_conflicting=True,
  File "/home/jens/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 826, in resolve
    dist = best[req.key] = env.best_match(req, ws, installer)
  File "/home/jens/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1098, in best_match
    return self.obtain(req, installer)
  File "/home/jens/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1110, in obtain
    return installer(requirement)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/dist.py", line 461, in fetch_build_egg
    return cmd.easy_install(req)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 664, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 694, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 875, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1114, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1100, in run_setup
    run_setup(setup_script, args)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 249, in run_setup
    raise
  File "/usr/lib64/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 197, in setup_context
    yield
  File "/usr/lib64/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 168, in save_modules
    saved_exc.resume()
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 143, in resume
    six.reraise(type, exc, self._tb)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 156, in save_modules
    yield saved
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 197, in setup_context
    yield
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 246, in run_setup
    DirectorySandbox(setup_dir).run(runner)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 276, in run
    return func()
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 245, in runner
    _execfile(setup_script, ns)
  File "/home/jens/.local/lib/python2.7/site-packages/setuptools/sandbox.py", line 47, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/setup.py", line 390, in <module>
    
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/setup.py", line 373, in setup_package
    
  File "/tmp/easy_install-IeB3Vc/numpy.git@numpysetup/setup.py", line 211, in generate_cython
    
RuntimeError: Running cythonize failed!
```

after running `easy_install --user cython`
and trying the above command again, installation works.

numpy's setup.py should declare a `setup_requires=['cython']` so I do not need to manually install cython before running install.

### PR DESCRIPTION
**Pull request by JensTimmerman**
fixes #7959

### CODE REVIEW

**Comment by charris** [comment]
Hmm, the debug test failed. That probably arises from line 62 in `tools/travis_test.sh`
```
    CFLAGS="$sysflags $werrors" $PYTHON setup.py build_ext --inplace
```
All the other tests except for `CHROOT` use `pip install .`. Not sure why that is a problem...

File: setup.py [hunk 1]
```diff
@@ -356,6 +356,7 @@ def setup_package():
         platforms = ["Windows", "Linux", "Solaris", "Mac OS-X", "Unix"],
         test_suite='nose.collector',
         cmdclass={"sdist": sdist_checked},
+        setup_requires=['cython'],
```
Comment (charris) [inline setup.py]: Does it also need `build_requires`?

**Comment by charris** [comment]
Hmm, `setup_requires` seems to be a strange beast. @rgommers Should we even attempt to use that option?

**Comment by pv** [comment]
setup_requires gets handled only when `setup()` runs. This is too late, because as it is currently the Cython .c files would be needed before running `setup()`. It's also a setuptools oddity which downloads stuff in a different way from pip etc., and probably best not relied on.

File: setup.py (hunk 1 above)
Comment (pv) [inline setup.py]: I think distutils/setuptools don't have a way of specifying build requirements?

**Comment by pv** [comment]
`setup_requires` is useful only if you have implemented things as setuptools command extensions that get called late enough inside the `setup()` call. matplotlib does things this way, but it's a lot of hoops to jump through. Of course, we have the numpy.distutils as an extra burden (matplotlib doesn't use it), and I don't really know (or want to know ;) how this complicates things.

**Comment by rgommers** [comment]
Cython is only a dependency for development builds not for releases (we ship all generated C files). Therefore we don't want Cython as a dependency in ``*_requires``.

Also, Cython should be treated as a compiler not a package that ``pip`` & co are allowed to upgrade. 

Plus, as @pv explained, there's really no way to install Cython in time anyway.

Therefore closing this PR.

**Pull request closed without merging**